### ISSUE TITLE
[Flags] enum does not serialize correctly

### ISSUE THREAD

**Issue by jacobdufault** [jacobdufault/fullserializer#65]
```c#
using UnityEngine;
using FullInspector;
using System;

public class Test2 : BaseBehavior {
    [Flags]
    public enum TestEnum {
        Option1 = 1 << 1,
        Option2 = 1 << 2,
        Option3 = 1 << 3,
    }

    public TestEnum enums;

    [InspectorButton]
    public void Print() {
        Debug.Log("Option1 = " + (int)TestEnum.Option1);
        Debug.Log("Option2 = " + (int)TestEnum.Option2);
        Debug.Log("Option3 = " + (int)TestEnum.Option3);
        Debug.Log("Option1|Option2 = " + (int)(TestEnum.Option1 | TestEnum.Option2));
        Debug.Log("Option1|Option2|Option3 = " + (int)(TestEnum.Option1 | TestEnum.Option2 | TestEnum.Option3));
    }
}
```

**Comment by jagt** [comment]
I noticed this a while ago. Just took some time to look into this and I believe it's a problem with FullInspector, as I can't reproduce this in FS.

**Comment by jacobdufault** [comment]
Hmm, thanks. I'll take a closer look.

I looked at the FI serialization dump and it looked like the values were not getting serialized properly. Though maybe the inspector was selecting the wrong values for serialization.

**Comment by helmesjo** [comment]
Any progress on this? Kinda stuck atm. :)

**Comment by jacobdufault** [comment]
Ok, I've done some digging. FS is serializing this correctly. There is a bug in Unity. I've submitted a report. You can work around the issue by adding an enum option with value zero, ie:

```c#
using UnityEngine;
using UnityEditor;
using System;

public class Error : MonoBehaviour {
}

[Flags]
public enum BrokenEnum {
    Option1 = 1 << 1,
    Option2 = 1 << 2,
    Option3 = 1 << 3,
}

[Flags]
public enum ValidEnum {
    Zero = 0, // This fixes the editor error.
    Option1 = 1 << 1,
    Option2 = 1 << 2,
    Option3 = 1 << 3,
}

[CustomEditor(typeof(Error))]
public class ErrorEditor : Editor {
    public override void OnInspectorGUI() {
        EditorGUILayout.EnumMaskField("Broken", BrokenEnum.Option1 | BrokenEnum.Option2);
        EditorGUILayout.EnumMaskField("Valid", ValidEnum.Option1 | ValidEnum.Option2);
    }
}
```

**Comment by helmesjo** [comment]
Actually, I already have a zero-value... But is the custom editor necessary for this bug? Only setting the Enum-attribute won't do it.

**Comment by helmesjo** [comment]
Any solution for this using a PropertyDrawer?

**Comment by jacobdufault** [comment]
Can you post the FI code that is demonstrating the bug with the zero-value enum? The code I posted above is just to reproduce the issue for Unity - apparently they are aware of it and (might) be fixing it.

**Comment by helmesjo** [comment]
Seems to be properties causing issues. Try the following:
```
public class enumtest : BaseBehavior {

	[Flags]
	public enum ValidEnum
	{
		Zero = 0, // This fixes the editor error.
		Option1 = 1 << 1,
		Option2 = 1 << 2,
		Option3 = 1 << 3,
	}

	public ValidEnum validEnum1 = ValidEnum.Zero;
	private ValidEnum validEnum2 = ValidEnum.Zero;

	[ShowInInspector]
	public ValidEnum validEnumProperty
	{
		get
		{
			return validEnum2;
		}
		set
		{
			validEnum2 = value;
		}
	}
}
```

**Comment by helmesjo** [comment]
Sidenote: My class in the actual project having issues is NOT a monobehaviour. It's just a plain serializable class and I noticed that the above ```ValidEnum``` did not work for it...

**Comment by lazlo-bonin** [comment]
There are two issues at hand here.

First, the EnumMaskField in Unity is bugged. It uses the order of the items instead of their actual bitmask values, meaning for example that any None value will left shift all the bit mask. I posted a fix for that on the Unity forums here: https://forum.unity.com/threads/editorguilayout-enummaskfield-doesnt-use-enums-values.233332/#post-3297123

Second, FullSerializer doesn't serialize compound flags to string correctly. It uses an OR instead of an AND check for compounds, meaning that for example an "All" value would always get serialized.

Basically, in `fsEnumConverter`, the following check has to be changed from:

    var isSet = (instanceValue & integralValue) != 0;

To:

    var isSet = (instanceValue & integralValue) == integralValue;

Which is consistent with the .NET Enum.HasFlag method.

Additionally, any None value always gets serialized, but this is useless and can be avoided, for example by adding:

    if (integralValue == 0) continue;
